rasterio/shim_rasterioex.pxi

```diff
@@ -34,13 +34,23 @@ cdef extern from "gdal.h" nogil:
     cdef CPLErr GDALDatasetRasterIOEx(GDALDatasetH hDS, GDALRWFlag eRWFlag, int nDSXOff, int nDSYOff, int nDSXSize, int nDSYSize, void *pBuffer, int nBXSize, int nBYSize, GDALDataType eBDataType, int nBandCount, int *panBandCount, GSpacing nPixelSpace, GSpacing nLineSpace, GSpacing nBandSpace, GDALRasterIOExtraArg *psExtraArg)
 
 
+gdal31_version_checked = False
+gdal31_version_met = False
 gdal33_version_checked = False
 gdal33_version_met = False
 
 def validate_resampling(resampling):
     """Validate that the resampling method is compatible of reads/writes"""
 
-    if resampling == Resampling.rms:
+    if resampling == Resampling.sum:
+        global gdal31_version_checked
+        global gdal31_version_met
+        if not gdal31_version_checked:
+            gdal31_version_checked = True
+            gdal31_version_met = GDALVersion.runtime().at_least('3.1')
+        if not gdal31_version_met:
+            raise ResamplingAlgorithmError("{!r} requires GDAL 3.1".format(Resampling(resampling)))
+    elif resampling == Resampling.rms:
```

Comment:
@visr lines 45-53 must be deleted. In GDAL versions before 3.3, only resampling values 0-6 are supported for read/write. Any other value will result in a segfault. In GDAL 3.3, there is a special case for `rms`: https://github.com/OSGeo/gdal/blob/master/gdal/gcore/gdal.h#L147. But there is no special case for `sum`. Does that make sense?